### PR TITLE
BGP support: extend project by BGP Config, new resource for BGP session

### PR DESCRIPTION
**Pull request by t0mk**
This PR will add support for BGP networking API resources as decribed in https://help.packet.net/article/56-bgp

Ready for review.

The usage is illustrated in the bgp_session resource doc.

### CODE REVIEW

File: packet/resource_packet_project.go [hunk 1]
```diff
@@ -57,10 +57,51 @@ func resourcePacketProject() *schema.Resource {
 				},
 				ValidateFunc: validation.StringMatch(uuidRE, "must be a valid UUID"),
 			},
+			"bgp_config": &schema.Schema{
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Computed: true,
+				ForceNew: false,
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_project.go]: Actually, update should be handled better for the `bgp_config`, I improved it, thanks for pointing it out!

File: packet/resource_packet_project.go [hunk 2]
```diff
@@ -57,10 +57,51 @@ func resourcePacketProject() *schema.Resource {
...
+			},
 		},
 	}
 }
 
+func expandBGPConfig(d *schema.ResourceData) packngo.CreateBGPConfigRequest {
+	bgpCreateRequest := packngo.CreateBGPConfigRequest{
+		DeploymentType: d.Get("bgp_config.0.deployment_type").(string),
+		Asn:            d.Get("bgp_config.0.asn").(int),
+		Md5:            d.Get("bgp_config.0.md5").(string),
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_project.go]: ok!

File: packet/resource_packet_project.go [hunk 3]
```diff
@@ -76,6 +117,25 @@ func resourcePacketProjectCreate(d *schema.ResourceData, meta interface{}) error
 
 	d.SetId(project.ID)
 
+	if _, ok := d.GetOk("bgp_config"); !ok {
+		d.Set("bgp_config", []interface{}{})
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_project.go]: No reason. Thanks for the comment!

File: packet/resource_packet_project.go [hunk 4]
```diff
@@ -96,16 +156,59 @@ func resourcePacketProjectRead(d *schema.ResourceData, meta interface{}) error {
...
-	d.Set("id", proj.ID)
+	d.SetId(proj.ID)
 	d.Set("payment_method_id", path.Base(proj.PaymentMethod.URL))
 	d.Set("name", proj.Name)
 	d.Set("organization_id", path.Base(proj.Organization.URL))
 	d.Set("created", proj.Created)
 	d.Set("updated", proj.Updated)
 
+	bgpConf, _, err := client.BGPConfig.Get(proj.ID, nil)
+	d.Set("bgp_config", []interface{}{})
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_project.go]: I thought that import test was failing without this, but now I tried and it goes through, so I removed it, thanks for pointing it out!

File: packet/resource_packet_bgp_session.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"device_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"address_family": {
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_bgp_session.go]: ok!

File: packet/resource_packet_project.go [hunk 6]
```diff
@@ -57,10 +57,51 @@ func resourcePacketProject() *schema.Resource {
...
 			},
+			"bgp_config": &schema.Schema{
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Computed: true,
+				ForceNew: false,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"deployment_type": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_project.go]: ok!

File: packet/resource_packet_bgp_session.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+	d.SetId(bgpSession.ID)
+	return resourcePacketBGPSessionRead(d, meta)
+}
+
+func resourcePacketBGPSessionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+	bgpSession, _, err := client.BGPSessions.Get(d.Id(),
+		&packngo.GetOptions{Includes: []string{"device"}})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (t0mk) [reply packet/resource_packet_bgp_session.go]: ok!

**Comment by t0mk** [comment]
@mbfrahry thanks for review! I fixed your comments. Biggest update is handling the `bgp_config` updated in the `packet_project` resource. Please take a look and once you give me a green light, I will merge.

File: packet/resource_packet_project.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbfrahry) [reply packet/resource_packet_project.go]: Minor nit: `ForceNew: false` is implied so we can remove that here

**Comment by t0mk** [comment]
@mbfrahry thanks! I removed the unnecessary forceNew. I will launch acceptance tests in Hashicorp Teamcity and if they pass, I will merge.

**Pull request merged**